8.0.2 types causing

## Describe the bug
Upgrading from 8.0.1 to 8.0.2 created an TypeScript:

```
app_1           | ./node_modules/next-i18next/src/appWithTranslation.tsx:61:3
app_1           | Type error: Type '((props: AppProps) => Element) & NonReactStatics<ComponentType<any>, {}>' is not assignable to type 'ComponentClass<P, any> | FunctionComponent<P> | (P extends SVGProps<SVGSymbolElement> ? "symbol" : never) | ... 173 more ... | (P extends SVGProps<...> ? "view" : never)'.
app_1           |   Type '((props: AppProps) => Element) & NonReactStatics<ComponentType<any>, {}>' is not assignable to type 'FunctionComponent<P>'.
app_1           |     Types of parameters 'props' and 'props' are incompatible.
app_1           |       Property 'pageProps' is missing in type 'Record<string, unknown> & { children?: ReactNode; }' but required in type 'AppProps'.
app_1           | 
app_1           |   59 |   }
app_1           |   60 | 
app_1           | > 61 |   return hoistNonReactStatics(AppWithTranslation, WrappedComponent)
app_1           |      |   ^
app_1           |   62 | }
app_1           |   63 | 
```

## Occurs in next-i18next version
When moving from 8.0.1 to 8.0.2 (likely the type inclusion caused it).

## Steps to reproduce
- Can't right now, can do over the weekend

## Expected behaviour
No error caused from the type casting.

## OS
 - Device: Linux, Ubuntu 20.10
 - Browser: N/A

I have also observed this, and other, type errors since upgrading to next-i18next@8.0.2:

```
$ tsc --noEmit

node_modules/next-i18next/src/appWithTranslation.tsx:18:3 - error TS2322: Type 'null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

18   configOverride: UserConfig = null,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/appWithTranslation.tsx:51:9 - error TS2322: Type 'i18n | null' is not assignable to type 'i18n'.
  Type 'null' is not assignable to type 'i18n'.

51         i18n={i18n}
           ~~~~

  node_modules/next-i18next/node_modules/react-i18next/src/index.d.ts:107:3
    107   i18n: i18n;
          ~~~~
    The expected type comes from property 'i18n' which is declared here on type 'IntrinsicAttributes & I18nextProviderProps & { children?: ReactNode; }'

node_modules/next-i18next/src/appWithTranslation.tsx:61:3 - error TS2322: Type '((props: AppProps) => Element) & NonReactStatics<ComponentType<any>, {}>' is not assignable to type 'ElementType<P>'.
  Type '((props: AppProps) => Element) & NonReactStatics<ComponentType<any>, {}>' is not assignable to type 'FunctionComponent<P>'.
    Types of parameters 'props' and 'props' are incompatible.
      Property 'pageProps' is missing in type 'Record<string, unknown> & { children?: ReactNode; }' but required in type 'AppProps'.

61   return hoistNonReactStatics(AppWithTranslation, WrappedComponent)
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  node_modules/next-i18next/src/appWithTranslation.tsx:13:3
    13   pageProps: SSRConfig
         ~~~~~~~~~
    'pageProps' is declared here.

node_modules/next-i18next/src/appWithTranslation.tsx:61:51 - error TS2345: Argument of type '"symbol" | "object" | "a" | "abbr" | "address" | "area" | "article" | "aside" | "audio" | "b" | "base" | "bdi" | "bdo" | "big" | "blockquote" | "body" | "br" | "button" | "canvas" | ... 159 more ... | FunctionComponent<...>' is not assignable to parameter of type 'ComponentType<any>'.
  Type '"symbol"' is not assignable to type 'ComponentType<any>'.

61   return hoistNonReactStatics(AppWithTranslation, WrappedComponent)
                                                     ~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:59:75 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type '(substring: string, ...args: any[]) => string'.
      Type 'undefined' is not assignable to type '(substring: string, ...args: any[]) => string'.

59         const defaultLocaleStructure = localeStructure.replace('{{lng}}', lng).replace('{{ns}}', defaultNS)
                                                                             ~~~

  node_modules/typescript/lib/lib.es5.d.ts:454:5
    454     replace(searchValue: string | RegExp, replacer: (substring: string, ...args: any[]) => string): string;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The last overload is declared here.

node_modules/next-i18next/src/config/createConfig.ts:80:34 - error TS7006: Parameter 'p' implicitly has an 'any' type.

80         const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
                                    ~

node_modules/next-i18next/src/config/createConfig.ts:80:61 - error TS7006: Parameter 'file' implicitly has an 'any' type.

80         const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
                                                               ~~~~

node_modules/next-i18next/src/config/createConfig.ts:110:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'UserConfig'.
  No index signature with a parameter of type 'string' was found on type 'UserConfig'.

110     if (userConfig[obj]) {
            ~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:111:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.

111       combinedConfig[obj] = {
          ~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:112:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultNS: string; errorStackTraceLimit: number; i18n: { defaultLocale: string; locales: string[]; }; readonly initImmediate: boolean; interpolation: { escapeValue: boolean; format: (value: string, format: string) => string; formatSeparator: string; }; ... 7 more ...; use: never[]; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultNS: string; errorStackTraceLimit: number; i18n: { defaultLocale: string; locales: string[]; }; readonly initImmediate: boolean; interpolation: { escapeValue: boolean; format: (value: string, format: string) => string; formatSeparator: string; }; ... 7 more ...; use: never[]; }'.

112         ...defaultConfig[obj],
               ~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:113:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.

113         ...combinedConfig[obj],
               ~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:114:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'UserConfig'.
  No index signature with a parameter of type 'string' was found on type 'UserConfig'.

114         ...userConfig[obj],
               ~~~~~~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:2:30 - error TS7016: Could not find a declaration file for module 'i18next-fs-backend/cjs'. '/Users/alex.macpherson/code/ku-search-fe/node_modules/i18next-fs-backend/cjs/index.js' implicitly has an 'any' type.

2 import i18nextFSBackend from 'i18next-fs-backend/cjs'
                               ~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:16:5 - error TS2532: Object is possibly 'undefined'.

16     config.use.forEach(x => instance.use(x))
       ~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:19:28 - error TS2454: Variable 'initPromise' is used before being assigned.

19   return { i18n: instance, initPromise }
                              ~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:14:3 - error TS2322: Type 'null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

14   configOverride: UserConfig = null,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:62:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.

62       initialI18nStore[locale][ns] = (
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:72:7 - error TS2322: Type 'UserConfig | null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type 'UserConfig'.
    Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

72       userConfig: config.serializeConfig ? userConfig : null,
         ~~~~~~~~~~

  node_modules/next-i18next/types.d.ts:53:5
    53     userConfig: UserConfig
           ~~~~~~~~~~
    The expected type comes from property 'userConfig' which is declared here on type '{ initialI18nStore: any; initialLocale: string; userConfig: UserConfig; }'


Found 18 errors.
```

@isaachinman My tsconfig contains both `"skipLibCheck": true` and `"exclude": ["node_modules"]` - same as @Cretezy.

I've just cloned the repo you posted and played around with the tsconfig settings to reproduce the issue. Only change needed is to set `strict` to be `true` and all the below errors are presented:

```
next-i18next-typescript % yarn tsc --noEmit
yarn run v1.22.4
warning package.json: No license field
$ /Users/alex.macpherson/code/next-i18next-typescript/node_modules/.bin/tsc --noEmit
components/Header.tsx:3:26 - error TS7031: Binding element 'title' implicitly has an 'any' type.

3 export const Header = ({ title }) => (
                           ~~~~~

node_modules/next-i18next/src/appWithTranslation.tsx:2:34 - error TS7016: Could not find a declaration file for module 'hoist-non-react-statics'. '/Users/alex.macpherson/code/next-i18next-typescript/node_modules/hoist-non-react-statics/dist/hoist-non-react-statics.cjs.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/hoist-non-react-statics` if it exists or add a new declaration (.d.ts) file containing `declare module 'hoist-non-react-statics';`

2 import hoistNonReactStatics from 'hoist-non-react-statics'
                                   ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/appWithTranslation.tsx:18:3 - error TS2322: Type 'null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

18   configOverride: UserConfig = null,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/appWithTranslation.tsx:51:9 - error TS2322: Type 'i18n | null' is not assignable to type 'i18n'.
  Type 'null' is not assignable to type 'i18n'.

51         i18n={i18n}
           ~~~~

  node_modules/react-i18next/src/ts4.1/index.d.ts:210:3
    210   i18n: i18n;
          ~~~~
    The expected type comes from property 'i18n' which is declared here on type 'IntrinsicAttributes & I18nextProviderProps & { children?: ReactNode; }'

node_modules/next-i18next/src/config/createConfig.ts:59:75 - error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type '(substring: string, ...args: any[]) => string'.
      Type 'undefined' is not assignable to type '(substring: string, ...args: any[]) => string'.

59         const defaultLocaleStructure = localeStructure.replace('{{lng}}', lng).replace('{{ns}}', defaultNS)
                                                                             ~~~

  node_modules/typescript/lib/lib.es5.d.ts:454:5
    454     replace(searchValue: string | RegExp, replacer: (substring: string, ...args: any[]) => string): string;
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The last overload is declared here.

node_modules/next-i18next/src/config/createConfig.ts:80:34 - error TS7006: Parameter 'p' implicitly has an 'any' type.

80         const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
                                    ~

node_modules/next-i18next/src/config/createConfig.ts:80:61 - error TS7006: Parameter 'file' implicitly has an 'any' type.

80         const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
                                                               ~~~~

node_modules/next-i18next/src/config/createConfig.ts:110:9 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'UserConfig'.
  No index signature with a parameter of type 'string' was found on type 'UserConfig'.

110     if (userConfig[obj]) {
            ~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:111:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.

111       combinedConfig[obj] = {
          ~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:112:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultNS: string; errorStackTraceLimit: number; i18n: { defaultLocale: string; locales: string[]; }; readonly initImmediate: boolean; interpolation: { escapeValue: boolean; format: (value: string, format: string) => string; formatSeparator: string; }; ... 7 more ...; use: never[]; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultNS: string; errorStackTraceLimit: number; i18n: { defaultLocale: string; locales: string[]; }; readonly initImmediate: boolean; interpolation: { escapeValue: boolean; format: (value: string, format: string) => string; formatSeparator: string; }; ... 7 more ...; use: never[]; }'.

112         ...defaultConfig[obj],
               ~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:113:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.
  No index signature with a parameter of type 'string' was found on type '{ defaultLocale: string; locales: string[]; localeExtension: string; localePath: string; localeStructure: string; serializeConfig: boolean; strictMode: boolean; use: any[] | never[]; debug?: boolean | undefined; ... 46 more ...; errorStackTraceLimit: number; }'.

113         ...combinedConfig[obj],
               ~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/config/createConfig.ts:114:12 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'UserConfig'.
  No index signature with a parameter of type 'string' was found on type 'UserConfig'.

114         ...userConfig[obj],
               ~~~~~~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:2:30 - error TS7016: Could not find a declaration file for module 'i18next-fs-backend/cjs'. '/Users/alex.macpherson/code/next-i18next-typescript/node_modules/i18next-fs-backend/cjs/index.js' implicitly has an 'any' type.

2 import i18nextFSBackend from 'i18next-fs-backend/cjs'
                               ~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:16:5 - error TS2532: Object is possibly 'undefined'.

16     config.use.forEach(x => instance.use(x))
       ~~~~~~~~~~

node_modules/next-i18next/src/createClient/node.ts:19:28 - error TS2454: Variable 'initPromise' is used before being assigned.

19   return { i18n: instance, initPromise }
                              ~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:14:3 - error TS2322: Type 'null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

14   configOverride: UserConfig = null,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:62:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{}'.
  No index signature with a parameter of type 'string' was found on type '{}'.

62       initialI18nStore[locale][ns] = (
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/next-i18next/src/serverSideTranslations.ts:72:7 - error TS2322: Type 'UserConfig | null' is not assignable to type 'UserConfig'.
  Type 'null' is not assignable to type 'UserConfig'.
    Type 'null' is not assignable to type '{ i18n: NextJsI18NConfig; localeExtension?: string | undefined; localePath?: string | undefined; localeStructure?: string | undefined; serializeConfig: boolean; strictMode?: boolean | undefined; use?: any[] | undefined; }'.

72       userConfig: config.serializeConfig ? userConfig : null,
         ~~~~~~~~~~

  node_modules/next-i18next/types.d.ts:53:5
    53     userConfig: UserConfig
           ~~~~~~~~~~
    The expected type comes from property 'userConfig' which is declared here on type '{ initialI18nStore: any; initialLocale: string; userConfig: UserConfig; }'

pages/_app.tsx:3:18 - error TS7031: Binding element 'Component' implicitly has an 'any' type.

3 const MyApp = ({ Component, pageProps }) => <Component {...pageProps} />
                   ~~~~~~~~~

pages/_app.tsx:3:29 - error TS7031: Binding element 'pageProps' implicitly has an 'any' type.

3 const MyApp = ({ Component, pageProps }) => <Component {...pageProps} />
                              ~~~~~~~~~

pages/index.tsx:42:40 - error TS7031: Binding element 'locale' implicitly has an 'any' type.

42 export const getStaticProps = async ({ locale }) => ({
                                          ~~~~~~

pages/second-page.tsx:30:40 - error TS7031: Binding element 'locale' implicitly has an 'any' type.

30 export const getStaticProps = async ({ locale }) => ({
                                          ~~~~~~


Found 22 errors.
```

Here's the offending tsconfig from that repo for clarity:

```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve"
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ],
  "exclude": [
    "node_modules"
  ]
}
```

That said, I'm not sure why my own project isn't adhering to the `skipLibCheck` rule and throwing these errors from `next-i18next@8.0.2` 🤔

--- a/package.json
+++ b/package.json
@@ -72,6 +72,7 @@
     "@babel/preset-react": "^7.10.4",
     "@babel/preset-typescript": "^7.10.4",
     "@testing-library/react": "^11.2.5",
+    "@types/i18next-fs-backend": "^1.0.0",
     "@types/jest": "^24.0.16",
     "@types/node": "^14.14.19",
     "@types/react": "^16.8.4",
@@ -101,6 +102,7 @@
     "typescript": "^4.1.3"
   },
   "dependencies": {
+    "@types/hoist-non-react-statics": "^3.3.1",
     "core-js": "^3",
     "hoist-non-react-statics": "^3.2.0",
     "i18next": "^19.8.4",

--- a/src/appWithTranslation.client.test.tsx
+++ b/src/appWithTranslation.client.test.tsx
@@ -30,7 +30,12 @@ const renderComponent = () =>
       pageProps={{
         _nextI18Next: {
           initialLocale: 'en',
-          userConfig: {},
+          userConfig: {
+            i18n: {
+              defaultLocale: 'en',
+              locales: ['en', 'de'],
+            },
+          },
         },
       } as any}
     />
@@ -54,6 +59,10 @@ describe('appWithTranslation', () => {
       <div>Hello world</div>
     ), {
       configOverride: 'custom-value',
+      i18n: {
+        defaultLocale: 'en',
+        locales: ['en', 'de'],
+      },
     } as any)
     render(
       <DummyAppConfigOverride

--- a/src/appWithTranslation.client.test.tsx
+++ b/src/appWithTranslation.client.test.tsx
@@ -10,7 +10,7 @@ jest.mock('fs', () => ({
   readdirSync: jest.fn(),
 }))
 
-const DummyI18nextProvider = ({ children }) => (
+const DummyI18nextProvider: React.FC = ({ children }) => (
   <>{children}</>
 )
 
@@ -32,7 +32,7 @@ const renderComponent = () =>
           initialLocale: 'en',
           userConfig: {},
         },
-      }}
+      } as any}
     />
   )
 
@@ -61,7 +61,7 @@ describe('appWithTranslation', () => {
           _nextI18Next: {
             initialLocale: 'en',
           },
-        }}
+        } as any}
       />
     )
     const [args] = (I18nextProvider as jest.Mock).mock.calls
@@ -82,7 +82,7 @@ describe('appWithTranslation', () => {
               initialLocale: 'en',
               userConfig: null,
             },
-          }}
+          } as any}
         />
       )
     ).toThrow('appWithTranslation was called without a next-i18next config')

--- a/src/appWithTranslation.server.test.tsx
+++ b/src/appWithTranslation.server.test.tsx
@@ -14,7 +14,7 @@ jest.mock('fs', () => ({
   readdirSync: jest.fn(),
 }))
 
-const DummyI18nextProvider = ({ children }) => (
+const DummyI18nextProvider: React.FC = ({ children }) => (
   <>{children}</>
 )
 
@@ -35,7 +35,7 @@ const renderComponent = () =>
         _nextI18Next: {
           initialLocale: 'en',
         },
-      }}
+      } as any}
     />,
   )
 

--- a/src/appWithTranslation.server.test.tsx
+++ b/src/appWithTranslation.server.test.tsx
@@ -34,6 +34,12 @@ const renderComponent = () =>
       pageProps={{
         _nextI18Next: {
           initialLocale: 'en',
+          userConfig: {
+            i18n: {
+              defaultLocale: 'en',
+              locales: ['en', 'fr'],
+            },
+          },
         },
       } as any}
     />,

--- a/src/appWithTranslation.tsx
+++ b/src/appWithTranslation.tsx
@@ -49,19 +49,22 @@ export const appWithTranslation = (
       }))
     }
 
-    if (i18n === null) {
-      throw new Error('next-i18next could not initialise an i18next client')
-    }
-
-    return (
-      <I18nextProvider
-        i18n={i18n}
-      >
-        <WrappedComponent
-          key={locale}
-          {...props}
-        />
-      </I18nextProvider>
+    return i18n !== null ? (
+      (
+        <I18nextProvider
+          i18n={i18n}
+        >
+          <WrappedComponent
+            key={locale}
+            {...props}
+          />
+        </I18nextProvider>
+      )
+    ) : (
+      <WrappedComponent
+        key={locale}
+        {...props}
+      />
     )
   }
 

--- a/src/appWithTranslation.tsx
+++ b/src/appWithTranslation.tsx
@@ -13,11 +13,10 @@ type AppProps = {
   pageProps: SSRConfig
 }
 
-export const appWithTranslation = <P extends Record<string, unknown>>(
-  WrappedComponent: React.ComponentType | React.ElementType,
-  configOverride: UserConfig = null,
-):
-  React.ComponentType<P> | React.ElementType<P> => {
+export const appWithTranslation = (
+  WrappedComponent: React.ComponentType<AppProps> | React.ElementType<AppProps>,
+  configOverride: UserConfig | null = null,
+) => {
   const AppWithTranslation = (props: AppProps) => {
     let i18n = null
     let locale = null
@@ -34,6 +33,10 @@ export const appWithTranslation = <P extends Record<string, unknown>>(
         userConfig = configOverride
       }
 
+      if (!userConfig?.i18n) {
+        throw new Error('appWithTranslation was called without config.i18n')
+      }
+
       locale = initialLocale;
 
       ({ i18n } = createClient({
@@ -44,6 +47,10 @@ export const appWithTranslation = <P extends Record<string, unknown>>(
         lng: initialLocale,
         resources: initialI18nStore,
       }))
+    }
+
+    if (i18n === null) {
+      throw new Error('next-i18next could not initialise an i18next client')
     }
 
     return (
@@ -58,5 +65,8 @@ export const appWithTranslation = <P extends Record<string, unknown>>(
     )
   }
 
-  return hoistNonReactStatics(AppWithTranslation, WrappedComponent)
+  return hoistNonReactStatics(
+    AppWithTranslation,
+    WrappedComponent,
+  )
 }

--- a/src/config/createConfig.test.ts
+++ b/src/config/createConfig.test.ts
@@ -9,15 +9,12 @@ jest.mock('fs', () => ({
   readdirSync: jest.fn(),
 }))
 
-const deleteWindow = () =>
-  global.window
-
 describe('createConfig', () => {
 
   describe('server side', () => {
     beforeAll(() => {
       Object.assign(process, { browser: false })
-      deleteWindow()
+      delete (global as any).window
     })
 
     describe('when filesystem is as expected', () => {

--- a/src/config/createConfig.test.ts
+++ b/src/config/createConfig.test.ts
@@ -9,12 +9,15 @@ jest.mock('fs', () => ({
   readdirSync: jest.fn(),
 }))
 
+const deleteWindow = () =>
+  global.window
+
 describe('createConfig', () => {
 
   describe('server side', () => {
     beforeAll(() => {
       Object.assign(process, { browser: false })
-      delete global.window
+      deleteWindow()
     })
 
     describe('when filesystem is as expected', () => {
@@ -106,7 +109,7 @@ describe('createConfig', () => {
 
         const config = createConfig.bind(null, {
           lng: 'en',
-        })
+        } as any)
 
         expect(config).toThrow('Default namespace not found at public/locales/en/common.json')
       })

--- a/src/config/createConfig.ts
+++ b/src/config/createConfig.ts
@@ -1,7 +1,7 @@
 import { defaultConfig } from './defaultConfig'
 import { InternalConfig, UserConfig } from '../../types'
 
-const deepMergeObjects = ['backend', 'detection']
+const deepMergeObjects = ['backend', 'detection'] as (keyof Pick<UserConfig, 'backend' | 'detection'>)[]
 
 export const createConfig = (userConfig: UserConfig): InternalConfig => {
   if (typeof userConfig?.lng !== 'string') {
@@ -55,7 +55,7 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
       // Validate defaultNS
       // https://github.com/isaachinman/next-i18next/issues/358
       //
-      if (typeof defaultNS === 'string') {
+      if (typeof defaultNS === 'string' && typeof lng !== 'undefined') {
         const defaultLocaleStructure = localeStructure.replace('{{lng}}', lng).replace('{{ns}}', defaultNS)
         const defaultFile = `/${defaultLocaleStructure}.${localeExtension}`
         const defaultNSPath = path.join(localePath, defaultFile)
@@ -77,7 +77,10 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
       // Set server side preload (namespaces)
       //
       if (!combinedConfig.ns) {
-        const getAllNamespaces = p => fs.readdirSync(p).map(file => file.replace(`.${localeExtension}`, ''))
+        const getAllNamespaces = (p: string) =>
+          fs.readdirSync(p).map(
+            (file: string) => file.replace(`.${localeExtension}`, '')
+          )
         combinedConfig.ns = getAllNamespaces(path.resolve(process.cwd(), `${serverLocalePath}/${lng}`))
       }
     }
@@ -109,7 +112,6 @@ export const createConfig = (userConfig: UserConfig): InternalConfig => {
   deepMergeObjects.forEach((obj) => {
     if (userConfig[obj]) {
       combinedConfig[obj] = {
-        ...defaultConfig[obj],
         ...combinedConfig[obj],
         ...userConfig[obj],
       }

--- a/src/createClient/browser.test.ts
+++ b/src/createClient/browser.test.ts
@@ -1,11 +1,10 @@
-import { InternalConfig } from '../../types'
 import createClientBrowser from './browser'
 
 const config = {
   defaultLocale: 'en',
   locales: ['en', 'de'],
   use: [],
-} as InternalConfig
+} as any
 
 describe('createClientBrowser', () => {
   it('returns a browser client', () => {

--- a/src/createClient/browser.ts
+++ b/src/createClient/browser.ts
@@ -7,8 +7,10 @@ export default (config: InternalConfig): CreateClientReturn => {
   let initPromise: InitPromise
 
   if (!instance.isInitialized) {
-    config.use.forEach(x => instance.use(x))
+    config?.use?.forEach(x => instance.use(x))
     initPromise = instance.init(config)
+  } else {
+    initPromise = Promise.resolve(i18n.t)
   }
 
   return { i18n: instance, initPromise }

--- a/src/createClient/node.test.ts
+++ b/src/createClient/node.test.ts
@@ -1,11 +1,10 @@
-import { InternalConfig } from '../../types'
 import createClientNode from './node'
 
 const config = {
   defaultLocale: 'en',
   locales: ['en', 'de'],
   use: [],
-} as InternalConfig
+} as any
 
 describe('createClientBrowser', () => {
   it('returns a node client', () => {

--- a/src/createClient/node.ts
+++ b/src/createClient/node.ts
@@ -1,5 +1,5 @@
 import i18n from 'i18next'
-import i18nextFSBackend from 'i18next-fs-backend/cjs'
+import i18nextFSBackend from 'i18next-fs-backend'
 
 import { InternalConfig, CreateClientReturn, InitPromise } from '../../types'
 
@@ -13,8 +13,11 @@ export default (config: InternalConfig): CreateClientReturn => {
       instance.use(i18nextFSBackend)
     }
 
-    config.use.forEach(x => instance.use(x))
+    config?.use?.forEach(x => instance.use(x))
     initPromise = instance.init(config)
+  } else {
+    initPromise = Promise.resolve(i18n.t)
   }
+
   return { i18n: instance, initPromise }
 }

--- a/src/serverSideTranslations.test.ts
+++ b/src/serverSideTranslations.test.ts
@@ -16,7 +16,7 @@ describe('serverSideTranslations', () => {
   afterEach(jest.resetAllMocks)
 
   it('throws if initialLocale is not passed', async () => {
-    await expect(serverSideTranslations(undefined))
+    await expect(serverSideTranslations(undefined as any))
       .rejects
       .toThrow('Initial locale argument was not passed into serverSideTranslations')
   })

--- a/src/serverSideTranslations.test.ts
+++ b/src/serverSideTranslations.test.ts
@@ -36,15 +36,25 @@ describe('serverSideTranslations', () => {
   })
 
   it('returns props', async () => {
-    const props = await serverSideTranslations('en')
+    const props = await serverSideTranslations('en-US', [], {
+      i18n: {
+        defaultLocale: 'en-US',
+        locales: ['en-US', 'fr-CA'],
+      },
+    } as UserConfig)
 
     expect(props).toEqual({
       _nextI18Next: {
         initialI18nStore: {
-          en: {},
+          'en-US': {},
         },
-        initialLocale: 'en',
-        userConfig: null,
+        initialLocale: 'en-US',
+        userConfig: {
+          i18n: {
+            defaultLocale: 'en-US',
+            locales: ['en-US', 'fr-CA'],
+          },
+        },
       },
     })
   })

--- a/src/serverSideTranslations.ts
+++ b/src/serverSideTranslations.ts
@@ -11,7 +11,7 @@ const DEFAULT_CONFIG_PATH = './next-i18next.config.js'
 export const serverSideTranslations = async (
   initialLocale: string,
   namespacesRequired: string[] = [],
-  configOverride: UserConfig = null,
+  configOverride: UserConfig | null = null,
 ): Promise<SSRConfig> => {
   if (typeof initialLocale !== 'string') {
     throw new Error('Initial locale argument was not passed into serverSideTranslations')
@@ -21,6 +21,10 @@ export const serverSideTranslations = async (
 
   if (fs.existsSync(path.resolve(DEFAULT_CONFIG_PATH))) {
     userConfig = await import(path.resolve(DEFAULT_CONFIG_PATH))
+  }
+
+  if (userConfig === null) {
+    throw new Error('next-i18next was unable to find a user config')
   }
 
   const config = createConfig({
@@ -41,7 +45,7 @@ export const serverSideTranslations = async (
 
   await initPromise
 
-  const initialI18nStore = {
+  const initialI18nStore: Record<string, any> = {
     [initialLocale]: {},
   }
 

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -2,6 +2,7 @@
   "compilerOptions": {
     "esModuleInterop": true,
     "noEmit": true,
+    "strict": true,
     "skipLibCheck": true,
     "lib": ["esnext", "dom"],
     "target": "es5",

--- a/types.d.ts
+++ b/types.d.ts
@@ -19,7 +19,7 @@ export type UserConfig = {
   localeExtension?: string
   localePath?: string
   localeStructure?: string
-  serializeConfig: boolean
+  serializeConfig?: boolean
   strictMode?: boolean
   use?: any[]
 } & InitOptions
@@ -50,7 +50,7 @@ export type SSRConfig = {
   _nextI18Next: {
     initialI18nStore: any
     initialLocale: string
-    userConfig: UserConfig
+    userConfig: UserConfig | null
   }
 }
 

